lib/identify.c

```diff
@@ -126,51 +131,132 @@ static void perspective_unmap(const double *c,
+			    int x0, int y0,
+			    int from, int to,
 			    span_func_t func, void *user_data,
-			    int depth)
+			    struct flood_fill_vars *stack,
+			    int stack_size)
 {
-	int left = x;
-	int right = x;
+	struct flood_fill_vars *vars;
+	struct flood_fill_vars *next_vars;
+	const struct flood_fill_vars *const last_vars = &stack[stack_size - 1];
 	int i;
-	quirc_pixel_t *row = q->pixels + y * q->w;
+	quirc_pixel_t *row;
 
-	if (depth >= FLOOD_FILL_MAX_DEPTH)
-		return;
+	/* Set up the first context  */
+	next_vars = &stack[0];
+	next_vars->x = x0;
+	next_vars->y = y0;
 
-	while (left > 0 && row[left - 1] == from)
-		left--;
+call:
+	vars = next_vars;
+	vars->left = vars->x;
+	vars->right = vars->x;
 
-	while (right < q->w - 1 && row[right + 1] == from)
-		right++;
+	row = q->pixels + vars->y * q->w;
+
+	while (vars->left > 0 && row[vars->left - 1] == from)
+		vars->left--;
+
+	while (vars->right < q->w - 1 && row[vars->right + 1] == from)
+		vars->right++;
 
 	/* Fill the extent */
-	for (i = left; i <= right; i++)
+	for (i = vars->left; i <= vars->right; i++)
 		row[i] = to;
 
 	if (func)
-		func(user_data, y, left, right);
+		func(user_data, vars->y, vars->left, vars->right);
 
-	/* Seed new flood-fills */
-	if (y > 0) {
-		row = q->pixels + (y - 1) * q->w;
+	if (vars == last_vars) {
+		return;
+	}
 
-		for (i = left; i <= right; i++)
-			if (row[i] == from)
-				flood_fill_seed(q, i, y - 1, from, to,
-						func, user_data, depth + 1);
+	/* Seed new flood-fills */
+	if (vars->y > 0) {
+		row = q->pixels + (vars->y - 1) * q->w;
+
+		for (i = vars->left; i <= vars->right; i++)
+			if (row[i] == from) {
+				/* Save the current context */
+				vars->i = i;
+				vars->pc = 1;
+
+				/* Set up the next context */
+				next_vars = vars + 1;
+				next_vars->x = i;
+				next_vars->y = vars->y - 1;
+				goto call;
+return_from_call1: ;
+			}
 	}
 
-	if (y < q->h - 1) {
-		row = q->pixels + (y + 1) * q->w;
+	if (vars->y < q->h - 1) {
+		row = q->pixels + (vars->y + 1) * q->w;
+
+		for (i = vars->left; i <= vars->right; i++)
+			if (row[i] == from) {
+				/* Save the current context */
+				vars->i = i;
+				vars->pc = 2;
+
+				/* Set up the next context */
+				next_vars = vars + 1;
+				next_vars->x = i;
+				next_vars->y = vars->y + 1;
+				goto call;
+return_from_call2: ;
+			}
+	}
 
-		for (i = left; i <= right; i++)
-			if (row[i] == from)
-				flood_fill_seed(q, i, y + 1, from, to,
-						func, user_data, depth + 1);
+	if (vars > stack) {
+		/* Restore the previous context */
+		vars--;
+		i = vars->i;
+		if (vars->pc == 1) {
+			row = q->pixels + (vars->y - 1) * q->w;
+			goto return_from_call1;
+		}
+		row = q->pixels + (vars->y + 1) * q->w;
+		goto return_from_call2;
 	}
 }
 
+static void flood_fill_prepare_workarea(const struct workarea *wa,
+					struct flood_fill_vars **stackp,
+					int *stack_sizep)
+{
+	/* ensure the alignment */
+
+	size_t align = sizeof(int); /* alignof(**stackp) */
```

Comment:
This doesn't seem to be a safe assumption to me, can you elaborate?

Reply:
the structure only contains several int members.
i'm not aware of any arch where this is not safe.